doc/manual/src/store/file-system-object/content-address.md

```diff
@@ -0,0 +1,83 @@
+
+For these reasons, Nix has its very own archive format—the Nix Archive (NAR) format,
+which is carefully designed to avoided the problems described above.
+
+The exact specification of the Nix Archive format is in `protocols/nix-archive.md`
+
+## Content addressing File System Objects beyond a single serialisation pass
+
+Serialising the entire tree and then hashing that binary string is not the only option for content addressing, however.
+There is also the [Merkle DAG](https://en.wikipedia.org/wiki/Merkle_tree) of multiple rounds of serialisation and hashing, where previously computed hashes are included in subsequent byte strings to be hashed.
+
+This creates the possibility of hashing reflecing the graph structure of file system objects:
+We can first hash (serialised) child objects, and then hash parent objects using the hashes of their children in the serialisation of the parent.
+With such a system, different file system object trees that share a common subtree will have the same hash for that common subtree, and this in turn opens the door to new methods of deduplication to save storage / bandwidth.
+Nix currently doesn't utilise these opportunities for deduplication, but that might change in the future.
+
+Currently, there is one such Merkle DAG content addressing method supported.
+
+### Git ([experimental][xp-feature-git-hashing]) { #git }
+
+> **Warning**
+>
+> This method is part of the [`git-hashing`][xp-feature-git-hashing] experimental feature.
+
+Git's file system model is very close to Nix's, and so Git's content addressing method is a pretty good fit.
+Just as with regular Git, files and symlinks are hashed as git "blobs", and directories are hashed as git "trees".
+
+However, this is one small difference between Nix's and Git's file system model that causes some friction.
+Plain files, executable files, and symlinks are not distinguished in and of themselves, but by their context: by the directory entry that refers to them.
+That means so long as the root object is a directory, there is no problem:
+every non-directory object is owned by a parent directory, and the entry that refers to it provides the missing information.
+However, if the root object is not a directory, then we have no way of knowing which one of an executable file, non-executable file, or symlink it is supposed to be.
+
+In response to this, we have decided to treat a bare file as non-executable file.
+This is similar to do what we do with [flat serialisation](#flat), which also lacks this information.
+To avoid an address collision, attempts to hash a bare executable file or symlink will result in an error (just as would happen for flat serialisation also).
+Thus, Git can encode some, but not all of Nix's "File System Objects", and this sort of content-addressing is likewise partial.
+
+In the future, we may adopt another Merkle DAG format which is capable of representing all Nix file system objects.
```

Comment:
What about hashing a directory with a `.` entry? That makes our things representable; just not directly usable in git.
It seems that we're promising not to support that, which maybe we shouldn't promise? Or should that be `git-nix` or something?

Reply:
```suggestion
In the future, we may support a Git-like hash for such file system objects, or we may adopt another Merkle DAG format which is capable of representing all Nix file system objects.
```